blogs/lokiless_netobserv/index.md

```diff
@@ -0,0 +1,126 @@
+
+    ![unfiltered topology view](images/owner_screenshot.png)
+
+* Cluster config: 3 worker and 3 master nodes, AWS m5.2xlarge machines
+* LokiStack size: 1x.extra-small
+
+Results:
+
+  Below table shows 90th Percentile query times for each table:
+
+  | Time Range | Loki      | Prometheus
+  | :--------: | :-------: | :----------:
+  | Last 5m    | 2287 ms   | 95.5 ms
+  | Last 1h    | 4581 ms   | 236 ms
+  | Last 6h    | > 10 s    | 394 ms
+
+As time range to fetch network flows gets wider, Loki queries tends to get slower or timing out, while Prometheus queries is able render the data within fraction of a second.
+
+### Resource utilization:
+In our tests conducted on 3 different test beds with varied workloads and network throughput, when Network Observability is configured without Loki, total savings of Memory usage could be in range 12-60% and CPU utilzation could be lower by 20-30%<sup>*</sup>. Not to mention you will not need to provision and plan for additional storage in public clouds for Loki, overall reducing the cost and improving operational efficiency significantly.
+
+In our perf tests, [kube-burner](https://github.com/kube-burner/kube-burner) workloads were used to generate several objects and create heavy network traffic. We used sampling rate of 1 for all below tests. To further describe each test bed:
+
+1. Test bed 1: node-density-heavy workload ran against 25 nodes cluster.
+2. Test bed 2: ingress-perf workload ran against 65 nodes cluster.
+3. Test bed 3: cluster-densit-v2 workload ran against 120 nodes cluster
+
+Below graphs shows total vCPU and memory usage for a recommended Network Observability stack (flowlogs-pipeline, eBPF-agent, Kafka and optionally Loki) for production clusters.
+
+![Compare total vCPUs utilized with and without Loki](<images/Total vCPUs consumed.png>)
+![Compare total RSS utilized with and without Loki](<images/Total Memory (RSS) consumed.png>)
+
+<sup>*</sup> actual resource utilization may depend on various factors such as flowcollector sampling size, number of workloads and nodes in an OCP cluster
+
+## Trade-offs:
+We saw having Prometheus as datasource provides impressive performance gains and sub-second query times, however it introduces below constraints:
+
+1. Without storage of network flows it no longer provides Traffic flows table. <TODO: insert a picture Traffic table greyed out>
+
+2. Per-pod level of resource granularity is not available since it causes Prometheus metrics to have high cardinality. <TODO: insert a picture where diff between with-Loki and without-Loki Scope>
```

Comment:
it would be interesting to also compare the output storage size with and without loki

Reply:
what do you mean by output storage size? I have mentioned here that users don't need to provision additional storage: 
https://github.com/netobserv/documents/pull/72/files#diff-77e07c919145d98ff5ecc61c81fe91d345f160619e8ddbf9e39d4dab956d7921R43

Reply:
Prometheus storage increases when enabling netobserv metrics.
There are even capabilities to export  to remote storages: https://prometheus.io/docs/prometheus/latest/storage/#remote-storage-integrations

The goal here would be to be able to say how much netobserv metrics could take (depending of what you enabled of course) compared to Loki storage when storing every flows

Reply:
> Prometheus storage increases when enabling netobserv metrics.

Do you think it would cause significant increase in prometheus storage by having these metrics published?

Reply:
It could depending of the cardinality of the selected metrics. But it's also important to highlight low impact configurations :smile_cat:

Reply:
@jotak do you have prometheus metrics to measure its storage util.? the ones on Prometheus dashboard for head_series and head_chunks - are those for prometheus in-memory storage for metrics? I did not find any pvc set for prometheus in OCP cluster.

Reply:
I found that estimation on grafana website:
```
The Prometheus metrics resource usage depends mainly on the number of active series that need to be scraped and the scrape interval.

As a rule of thumb, per each 1 million active series and with the default scrape interval, you can expect to use approximately:

0.4 CPU cores
11 GiB of memory
1.5 MiB/s of total network bandwidth, send and receive
```

https://grafana.com/docs/agent/latest/flow/tasks/estimate-resource-usage/#prometheus-metrics

Reply:
@jotak @jpinsonneau I did some testing to figure prometheus impact by:
```
when netobserv is deployed:
head series/chunks - jumps from ~470K to ~1M

Storage: `prometheus_tsdb_wal_storage_size_bytes` which shows high rate of increase [1] (not sure though if this is metric is indicator of prometheus' storage
Prom. CPU: doubles from 0.15 to 0.3 [2]
Prom. RSS:  jumps from 2.5G to 4.7 [3]
```
[1]
<img width="1201" alt="prom_wal_rate" src="https://github.com/user-attachments/assets/36be0632-843e-4682-990a-a1442809a805">

[2]
<img width="1192" alt="prom_cpu_usage" src="https://github.com/user-attachments/assets/faa5a3d8-ff64-4235-8f2f-a2121f25a2b5">

[3]
<img width="1187" alt="prom_memory_usage" src="https://github.com/user-attachments/assets/30b7a143-dd8c-4be3-8d8e-00771dd8abd9">


This is when default metrics are published when loki is disabled. 
I am not sure though this is the absolute difference between with and without Loki since many of the metrics we already published even when Loki is enabled.

Reply:
I think to see the total storage size you should use `prometheus_tsdb_storage_blocks_bytes: The number of bytes that are currently used for local storage by all blocks`.

You can also have a look to `prometheus_tsdb_head_chunks_storage_size_bytes: Size of the chunks_head directory` in parallel.

Anyways these results are not bad but we should inform the customers about these in the doc.

Reply:
@memodi when you say:
> Prom. CPU: doubles from 0.15 to 0.3 [2]
> Prom. RSS:  jumps from 2.5G to 4.7 [3]

Is it before, after, or during the NDH workloads being generated?

Reply:
@jotak it's during the NDH runs. CPU returns to the previous levels after the workload is completed, however RSS jump is more stickier probably because of those active series are still held in prometheus memory.

@jpinsonneau for some reason `prometheus_tsdb_storage_blocks_bytes` is always 0 in default config.

Reply:
@memodi so maybe it's also worth check what's the memory overhead if we run NDH without netobserv at all, because creating pods etc. does have an impact on metrics even without netobserv